spec/models/miq_expression_spec.rb

```diff
@@ -1631,4 +1631,146 @@
       end
     end
   end
+
+  describe "#sql_supports_atom?" do
+    context "expression key is 'CONTAINS'" do
+      context "operations with 'tag'" do
+        it "returns true if format of field is /\w+\.\w+/ - interpreted as tag of the main model" do
+          # example of expression:
+          # expression = {"CONTAINS" => {"tag" => "VmInfra.managed-operations", "value" => "analysis_failed"}}
+          expression = {"CONTAINS" => {"tag" => "HelloWorld.aaa"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(true)
+        end
+
+        it "returns false if format of field is /\w+\.\w+(\.\w+)+/ - interpreted as tag of not main model" do
+          # example of expression:
+          # field: "Vm.ext_management_system.managed-openshiftroles"
+          # expression = {"CONTAINS" => {"tag" => field, "value" => "node"}}
+          expression = {"CONTAINS" => {"tag" => "Hello.World.aaa"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(false)
+        end
+      end
+
+      context "operation with 'field'" do
+        it "returns false if format of field is /\w+(\.\w+)+/ - as model.association..association-field" do
+          # example of expression:
+          # field = "ManageIQ::Providers::InfraManager::Vm.service.user.vms-active"
+          # expression = {"CONTAINS" => {"field" => field, "value" => "true"}
+          expression = {"CONTAINS" => {"tag" => "Hello.World.aaa"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(false)
+        end
+
+        it "returns false if field belongs to virtual_has_many association" do
+          field = "ManageIQ::Providers::InfraManager::Vm.file_shares-type"
+          expression = {"CONTAINS" => {"field" => field, "value" => "abc"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(false)
+        end
+
+        it "returns false if field belongs to 'has_and_belongs_to_many' association" do
+          field = "ManageIQ::Providers::InfraManager::Vm.storages-name"
+          expression = {"CONTAINS" => {"field" => field, "value" => "abc"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(false)
+        end
+
+        it "returns false if field belongs to 'has_many' polimorhic association" do
+          field = "ManageIQ::Providers::InfraManager::Vm.advanced_settings-region_number"
+          expression = {"CONTAINS" => {"field" => field, "value" => "1"}}
+          expect(described_class.new(nil).sql_supports_atom?(expression)).to eq(false)
+        end
+
+        it "returns true if field belongs to 'has_many' association" do
```

Comment:
Have you covered these scenarios? https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_expression.rb#L871-L872